test/integration/Utils.kt

```diff
@@ -45,152 +46,194 @@ import java.nio.file.Path
+        wait(composeRule, 1_500)
 
-    composeRule.onNodeWithText(PLAY_ICON_STRING).performClick()
-    wait(composeRule, 2_500)
+        composeRule.onNodeWithText(CHECK_ICON_STRING).performClick()
+        wait(composeRule, 1_500)
 
-    composeRule.onNodeWithText(CHECK_ICON_STRING).performClick()
-    wait(composeRule, 500)
+        retry(composeRule, {StudioState.notification.queue.last().code == "CNX10"}, {}, FAIL_DATA_WRITE)
 
-    StudioState.client.session.close()
-}
+        StudioState.client.session.close()
+    }
 
-suspend fun verifyDataWrite(composeRule: ComposeContentTestRule, dbName: String, queryFileName: String) {
-    val queryString = fileNameToString(queryFileName)
-
-    composeRule.onNodeWithText("infer").performClick()
-    composeRule.waitForIdle()
-    composeRule.onNodeWithText("read").performClick()
-    wait(composeRule, 1_000)
-
-    TypeDB.coreClient(DB_ADDRESS).use { client ->
-        client.session(dbName, TypeDBSession.Type.DATA, TypeDBOptions.core().infer(true)).use { session ->
-            session.transaction(TypeDBTransaction.Type.READ).use { transaction ->
-                val results = ArrayList<String>()
-                val query = TypeQL.parseQuery<TypeQLMatch>(queryString)
-                transaction.query().match(query).forEach { result ->
-                    results.add(
-                        result.get("user-name").asAttribute().value.toString()
-                    )
+    suspend fun verifyDataWrite(composeRule: ComposeContentTestRule, address: String, dbName: String, queryFileName: String) {
+        val queryString = fileNameToString(queryFileName)
+
+        composeRule.onNodeWithText("infer").performClick()
+        composeRule.waitForIdle()
+        composeRule.onNodeWithText("read").performClick()
+        wait(composeRule, 1_000)
+
+        TypeDB.coreClient(address).use { client ->
+            client.session(dbName, TypeDBSession.Type.DATA, TypeDBOptions.core().infer(true)).use { session ->
+                session.transaction(TypeDBTransaction.Type.READ).use { transaction ->
+                    val results = ArrayList<String>()
+                    val query = TypeQL.parseQuery<TypeQLMatch>(queryString)
+                    transaction.query().match(query).forEach { result ->
+                        results.add(
+                            result.get("user-name").asAttribute().value.toString()
+                        )
+                    }
+                    assertEquals(2, results.size)
+                    assertTrue(results.contains("jmsfltchr"))
+                    assertTrue(results.contains("krishnangovindraj"))
                 }
-                assertEquals(2, results.size)
-                assertTrue(results.contains("jmsfltchr"))
-                assertTrue(results.contains("krishnangovindraj"))
             }
         }
     }
+
+    private suspend fun retry(composeRule: ComposeContentTestRule, successCondition: () -> Boolean,
```

Comment:
I'm interested on your feedback on this function in particular.

Reply:
I think we have a few issues here:

1. Name: What does the function do? It verifies a condition and then delays with a retry loop. This is not encapsulated in the name "retry", which is too broad.
2. Signature: Parameters should come in order of "relevance" and "importance". In this case, the function is pretty much _defined_ by its `successCondition`, so that should be first. `failMessage` affects the output but is an edge case (error case) so it should be later on (as it currently is). `composeRule` (correct me if I'm wrong) is like the function's "context", the background in which it applies. We typically want to provide that last of all.
3. Signature: `intermediate` is a very odd name for a parameter, that provides little knowledge on what it is.
4. Retry strategy: Using a fixed deadline (10 seconds) and fixed retry interval (0.5 seconds) seems a little odd here. Are all of the things we "retry" a similar "difficulty"? e.g: do we expect committing a schema to take the same length of time as connecting to the DB? It's possible that we do, in which case this is fine. But I think we need to communicate that somehow, be it the method name, PR comment or code comment.
5. Initial delay: We verify the condition at the very beginning of the loop with no delay whatsoever. Don't we always expect the condition to fail at that time? Perhaps we should delay before verifying, as opposed to after.

Reply:
1. Agreed. I'm tending towards `delayUntilTrue` with `successCondition` as the first argument, but it makes the inclusion of composeRule a little odd.
2. Understood, see previous comment.
3. Sometimes we have to take intermediate steps before we wait again, e.g. `StudioState.client.refreshDatabaseList()`. This is difficult to name.
4. My thinking is that 10 seconds is the maximum length of time anything should take, at which point we fail. Varying our deadline and retry intervals doesn't get us much (there are a lot of reasons a file write or a database connection might take much longer than usual) and means we have to include another parameter for our function. Would like to discuss this more.
5. The approach right now is that we wait how long we expect to wait before the retry is attempted and then enter the retry. This way, reading just the test is sufficient to establish the fact that we are delaying without having to know how this function works.

Reply:
Summary of an in-person conversation:
We've landed on waitForCondition as a name for the function itself.
Additionally, we prefer `beforeRetry` to `intermediate`. 
We're in agreement on my points 4 and 5.

Reply:
This has been implemented.